docs/container.md

```diff
@@ -0,0 +1,313 @@
+
+  ```php
+  use Piwik\Translation\Translator;
+  use Piwik\Plugins\MyPlugin\Dao\MyEntityDao;
+
+  class API
+  {
+      /**
+       * @var Translator
+       */
+      private $translator;
+
+      /**
+       * @var MyEntityDao
+       */
+      private $myEntityDao;
+
+      public function __construct(Translator $translator, MyEntityDao $myEntityDao)
+      {
+          $this->translator = $translator;
+          $this->myEntityDao = $myEntityDao;
+      }
+
+      // ...
+  }
+  ```
+
+3. **Determine if your class has any non-class dependencies.**
+
+  Sometimes your class may depend on data that is in the DI container and not just classes, for example, [path.tmp](https://github.com/piwik/piwik/blob/master/config/global.php).
+
+  If your class has such a dependency, add it to the constructor without a type and then manually configure the parameter in DI config. _See the next section to learn how to create DI config._
+
+  For example:
+
+  ```php
+  use Piwik\Translation\Translator;
+  use Piwik\Plugins\MyPlugin\Dao\MyEntityDao;
+
+  class API
+  {
+      /**
+       * @var Translator
+       */
+      private $translator;
+
+      /**
+       * @var MyEntityDao
+       */
+      private $myEntityDao;
+
+      /**
+       * @var string
+       */
+      private $tmpPath;
+
+      public function __construct(Translator $translator, MyEntityDao $myEntityDao, $tmpPath)
+      {
+          $this->translator = $translator;
+          $this->myEntityDao = $myEntityDao;
+          $this->tmpPath = $tmpPath;
+      }
+
+      // ...
+  }
+  ```
+
+* **Determine how your new class will be used and ensure it is constructed by the DI container.**
+
+  **If another one of your classes creates your new class...**
+
+  ... simply add the new class to the constructor of the class that's using it. If the class that's using it is stored in the container, then your new class will be constructed using DI automatically.
+
+  If the user class cannot be stored in the container, then you must use `Piwik\Container\StaticContainer::get('Piwik\Plugins\MyPlugin\MyNewClass')` to get the instance in the user class.
+
+  **If Piwik creates your new class...**
+
+  Some classes are constructed by Piwik core (for example, API classes, Report classes, etc.). Some of these classes will automatically be constructed using the DI container. For these classes, you don't have any extra work to do.
+
+  If the class is not constructed with DI, then you must use `StaticContainer` as described above.
+
+  The list of classes that are automatically constructed using DI include:
+
+  * API classes.
+  * Implementations of the [Auth interface](http://developer.piwik.org/api-reference/Piwik/Auth).
+  * `Widgets` classes.
+  * Plugin `Settings` classes.
+  * `Type` classes.
+  * Plugin `Menu` classes.
+  * Plugin `Tasks` classes.
+
+  In 3.0, all classes will be constructed with dependency injection.
+
+### Configuring the container
+
```

Comment:
I think for plugin developers relevant is 

* How to configure the DI container and how to replace dependencies etc (covered by this section)
* How to get an instance from DI container in the plugin itself (eg via constructor). Maybe we could also explain why one something has to use `StaticContainer` etc. Covered partially by the other sections but too detailed and too complicated I'd say. Maybe we can use that example `API` above to explain it.

Plugins should be pretty small usually and DI might be not too important for them. Eg they possibly don't need to create classes that can be changed by other plugins etc. This would be rather a use case for the core. I think for plugins more important is how can I change dependencies defined in core and how can I get the correct dependency eg for `Config`, `DB` in core